src/v/utils/named_type.h

```diff
@@ -144,39 +122,39 @@ class base_named_type<T, Tag, std::false_type> {
 
     base_named_type& operator=(const base_named_type& o) = default;
 
-    bool operator==(const base_named_type& other) const {
-        return _value == other._value;
-    }
-    bool operator!=(const base_named_type& other) const {
-        return _value != other._value;
-    }
-    bool operator<(const base_named_type& other) const {
-        return _value < other._value;
-    }
-    bool operator>(const base_named_type& other) const {
-        return _value > other._value;
-    }
-    bool operator<=(const base_named_type& other) const {
-        return _value <= other._value;
+    // provide overloads for naked type
+    bool operator==(const type& other) const noexcept {
+        return _value == other;
     }
-    bool operator>=(const base_named_type& other) const {
-        return _value >= other._value;
+
+    auto operator<=>(const type& other) const noexcept -> std::strong_ordering {
+        // roundabout way to cope with type when it does not provide <=>
+        if constexpr (std::three_way_comparable_with<type, type>)
+            return _value <=> other;
+        else {
+            if (_value == other) {
+                return std::strong_ordering::equal;
+            } else if (_value < other) {
+                return std::strong_ordering::less;
+            } else {
+                return std::strong_ordering::greater;
+            }
+        }
     }
 
-    // provide overloads for naked type
-    bool operator==(const type& other) const { return _value == other; }
-    bool operator!=(const type& other) const { return _value != other; }
-    bool operator<(const type& other) const { return _value < other; }
-    bool operator>(const type& other) const { return _value > other; }
-    bool operator<=(const type& other) const { return _value <= other; }
-    bool operator>=(const type& other) const { return _value >= other; }
+    // these can be constexpr in c++23
+    bool operator==(const base_named_type& o) const = default;
+    auto operator<=>(const base_named_type& o) const noexcept {
+        // delegate to <=> type version
+        return (*this) <=> o._value;
+    }
```

Comment:
Prefer hidden friends (and add the `noexcept`)
```suggestion
    friend bool
    operator==(const base_named_type& l, const base_named_type& r) noexcept
      = default;
    friend auto
    operator<=>(const base_named_type& l, const base_named_type& r) noexcept {
        return l <=> r._value;
    }
```
Which means these will now succeeed.
```c++
using named_sstring = named_type<ss::sstring, struct named_string_tag>;

static_assert(
  std::equality_comparable_with<
    std::reference_wrapper<named_sstring>,
    std::reference_wrapper<named_sstring>>);

static_assert(
  std::equality_comparable_with<
    named_sstring,
    std::reference_wrapper<named_sstring>>);

static_assert(
  std::equality_comparable_with<
    std::reference_wrapper<named_sstring>,
    named_sstring>);
```

Reply:
the hidden friend: i had a follow up that makes use of operator==|<=> being a member function (basically the driving reason to make this pr), but the case is compelling to make the reference_wrapper case succeed.

noexcept: i left it to default to add it but i can be explicit

Reply:
changed to friend